components/nvs/include/nvs.hpp

```diff
@@ -0,0 +1,194 @@
+#pragma once
+
+#include <string>
+
+#include "nvs.h"
+#include "nvs_flash.h"
+#include "nvs_handle.hpp"
+
+#include "base_component.hpp"
+
+namespace espp {
+/**
+ * @brief Class to interface with the NVS.
+ * @details This class is used to interface with the ESP NVS system. It is used to init and erase the partition, or get and set variables
+ *
+ * @section nvs_ex1 NVS Example
+ * @snippet nvs_example.cpp nvs example
+ */
+class Nvs : public BaseComponent {
+public:
+
+    /**
+     * @brief Construct a new NVS object.
+     */
+    explicit Nvs()
+      : BaseComponent( "Nvs", espp::Logger::Verbosity::WARN) {}
+
+    /// @brief Initialize the NVS
+    /// @param[out] ec Saves a std::error_code representing success or failure
+    /// @details If the flash init fails because of no free pages or new version, it erases the NVS and re-initializes
+    void init(std::error_code &ec) {
```

Comment:
It seems like either 
1. init shouldn't be a public function, since you're calling it within each of the other functions anyway, or
2. You shouldn't be calling init from within each of the other functions (since presumably they'll return the appropriate error if NVS wasn't initialized and it's just extra code if the user has already called `init()`)

I think my preference is for option (2), but I've not looked into the internals of `nvs_flash_init()` to see how quickly it returns if it was already initialized and how well it's protected against multiple threads calling it at the same time. If it's a quick return and it's protected anyway, then option (1) would be good as well.

Reply:
Just checked their code and it seems that
* They do protect against multiple threads calling it simultaneously (locking)
* It is not a trivial return if it is initialized since it goes through their partition manager and it does some lookup and such, so I'd suggest we go with option (2) and remove the init call from each of the other methods.